### ISSUE TITLE
User-specified inchannels not honored in XYZAlgorithm.py

### ISSUE THREAD

**Issue by erigler-usgs** [usgs/geomag-algorithms#205]
https://github.com/usgs/geomag-algorithms/blob/f5338dfbcb2e62ae5ca67aed886d65913cc80675/geomagio/algorithm/XYZAlgorithm.py#L137

Is there a compelling reason to hard-code `_inchannels` like at the link above? This can break things if one specifies the command line option `--inchannels` as something different than what is defined in `CHANNELS` at the beginning of XYZAlgorithm.py.

A specific example is this:

0) user wants to convert obs data into geo;
1) the input F channel is all gaps, but HEZ is fine;
2) TimeseriesUtility.get_merged_gaps() causes Algorithm.can_produce_data() to return False, which...
3) prevents Controller.py's run_as_update() to not process and push results to the output factory.
4) NOTE: this is only an issue when geomag.py's `--update` option is used.

However, if the user specifies `--inchannels` to be only HEZ, or even HE, an error is produced by XYZAlgorithm.check_stream(), even though neither Z or F are required for the transform.

If a sanity check is required/desired, perhaps `CHANNELS` could be modified to define only those channels that are specific to a particular transform. E.g., 

```
CHANNELS = {
    'geo': ['X', 'Y'],
    'mag': ['H', 'D'],
    'obs': ['H', 'E'],
    'obsd': ['H', 'D']
}
```

**Comment by jmfee-usgs** [comment]
In some cases the algorithm is used to translate HEZF to the full set of XYZF channels (an entire IAGA file at a time), but the algorithm could be updated to pass those through when they're available and not consider them during gap checking.  This will require a little more thought to avoid breaking the existing usage.

**Comment by jmfee-usgs** [comment]
I think there are a few problems

1) check_stream currently checks that all _inchannels exist:

The XYZAlgorithm only needs to check the first 2 channels in _inchannels.


2) the Controller calls can_produce_data when running in update mode:
https://github.com/usgs/geomag-algorithms/blob/master/geomagio/Controller.py#L242

I added a channels parameter to TimeseriesUtility.get_stream_gaps, which can be used to only check for gaps in specific channels:
https://github.com/usgs/geomag-algorithms/pull/208/files#diff-dfa221a4ba339c02f1a03f382e0fc4d0R7

The XYZAlgorithm only needs to check the first 2 channels in _inchannels.

> As an aside, the can_produce_data method of the default Algorithm requires all channels to be present.  When running the controller in update mode, this prevents any data from being generated.  The default Algorithm _could_ be modified to always return True (since it can return any data it received), but there would need to be a separate "requires all channels" type of base class other algorithms could use.


3) StreamConverter methods expect at least one trace for each channel
https://github.com/usgs/geomag-algorithms/blob/master/geomagio/StreamConverter.py#L31

These need to check Z/F channels more carefully, and pass along any that are found.

**Comment by arigdon-usgs** [comment]
The original check stream does not care if the entire trace is empty just that the trace exists. So  instead of returning an exception and breaking, what if the XYZAlgorithm created the missing Z and F channels as empty traces of NaNs. Then before returning the out_stream it removes whichever empty trace was added. This would mean that nothing would have to be changed in the StreamConverter or the Controller and the specified inchannels would also be the outchannels.

**Comment by jmfee-usgs** [comment]
I would try a custom can_produce_data method on the XYZAlgorithm first, and see if that is all that is needed (it is needed in any case).

It's possible that check_stream is not the problem, since IAGAFactory and EDGEFactory should fill in missing channels.  However, `can_produce_data` and `get_stream_gaps` definitely prevent processing.

**Comment by erigler-usgs** [comment]
There is still the problem with the functions in StreamConverter.py. They are all hard-coded to attempt to extract (or reference) all the originally specified channels (e.g.,

```python
x = geo.select(channel='X')[0]
y = geo.select(channel='Y')[0]
z = geo.select(channel='Z')[0]
fs = geo.select(channel='F')[0]
```
)

If those channels don't exist, crash! (more specifically, "IndexError: list index out of range".

So I think something like Adren's proposal is going to be necessary. I imagine you would implement this via the XYZAlgorithm.check_stream() method. Does that sound right @arigdon-usgs?

**Comment by jmfee-usgs** [comment]
I was suggesting that the edge factory probably creates those empty traces, full of numpy.nan's; however, the can_produce_data method sees they are gaps and stops processing.

**Comment by jmfee-usgs** [comment]
Fixed in #209
